### PR TITLE
[3.x] Fix error when having BoneAttachment before PhysicalBone

### LINKED ISSUE #42010: Having BoneAttachment node before a Physical Bone for the same bone throws error
**Godot version:** 3.2.2 stable

**OS/device including version:** Windows 10

**Issue description:**

having a bone attachment node attached to a non root bone before a physical bone for that same bone in the heirarchy as shown:
image

throws this error message:
 scene/3d/skeleton.cpp:637 - Condition "bones[p_bone].physical_bone" is true.
 ./core/list.h:158 - Condition "!p_I" is true. Returned: false
image

and this error message when run:
E 0:00:00.501   erase: Condition "!p_I" is true. Returned: false
  <C++ Source>  ./core/list.h:158 @ erase()
everything works fine, but you get that error message for each bone on scene load 

But if you rearrange it this way it works fine with no error message:
image

**Minimal reproduction project:**

[boneerrormessagedemonstration.zip](https://github.com/godotengine/godot/files/5213191/boneerrormessagedemonstration.zip)
project contains two scenes: HasErrorMessage.tscn and NoErrorMessage.tscn, which show it not working correctly and it working correctly respectively

### PR DESCRIPTION
**Pull request by timothyqiu**
Fixes #42010

`bind_child_node_to_bone()` allows re-adding an existing bind. But `unbind_child_node_from_bone()` does not have extra checks for unbinding a non-existing bindings. So `List::erase()` complains about erasing something that does not exist.

There's no error when no `BoneAttachment` node comes before a `PhysicalBone`. This is because calling `List::erase()` on an empty list won't print an error. (Bone attachments are also bind to bones, so when there's a bone attachment node before the physical bone, the bind node list is not empty when it enters tree.)

It's inconsistent that `List::erase()` behaves differently when empty and non-empty. But I think fixing this on `Skeleton` level is acceptable since we prefer local solutions :stuck_out_tongue: 

There're no such methods on `master`.

### CODE REVIEW

File: scene/3d/skeleton.cpp [hunk 1]
```diff
@@ -538,6 +536,11 @@ void Skeleton::unbind_child_node_from_bone(int p_bone, Node *p_node) {
 	ERR_FAIL_INDEX(p_bone, bones.size());
 
 	uint32_t id = p_node->get_instance_id();
+
+	if (bones.write[p_bone].nodes_bound.find(id) == nullptr) {
+		return; // Doesn't exist in the first place.
+	}
+
```
Comment (lawnjelly) [inline scene/3d/skeleton.cpp:543]: Can you store the element returned and use this to call `erase` using the element directly rather than by value?

**Comment by lawnjelly** [comment]
Might be worth reviewing at the same time whether this should be a `List` at all versus a `LocalVector`. There's a whole bunch of stuff in Godot using `List`s that really should be `LocalVector`s (due to linked lists being so bad for cache etc). At some point it would be nice to go through these.

**Comment by timothyqiu** [comment]
Updated and replaced the `List` with `LocalVectori`.

File: scene/3d/skeleton.cpp [hunk 2]
```diff
@@ -290,8 +291,9 @@ void Skeleton::_notification(int p_what) {
 					b.global_pose_override_amount = 0.0;
 				}
 
-				for (List<uint32_t>::Element *E = b.nodes_bound.front(); E; E = E->next()) {
-					Object *obj = ObjectDB::get_instance(E->get());
+				const LocalVectori<unsigned int> &nodes = b.nodes_bound;
```
Comment (lawnjelly) [inline scene/3d/skeleton.cpp:294]: Should this be `const LocalVectori<uint32_t> &nodes`? (and the one above?)

File: scene/3d/skeleton.cpp [hunk 3]
```diff
@@ -538,13 +538,20 @@ void Skeleton::unbind_child_node_from_bone(int p_bone, Node *p_node) {
...
 
 	uint32_t id = p_node->get_instance_id();
-	bones.write[p_bone].nodes_bound.erase(id);
+
+	int index = bones[p_bone].nodes_bound.find(id);
+	if (index == -1) {
+		return; // Doesn't exist in the first place.
+	}
+
+	bones.write[p_bone].nodes_bound.remove(index);
```
Comment (lawnjelly) [inline scene/3d/skeleton.cpp:547]: Are the `nodes_bound` required to be in order? If not, we can maybe use `remove_unordered()`.. :thinking:

Comment (timothyqiu) [reply scene/3d/skeleton.cpp:547]: `Skeleton` itself does not require some specific order. But changing `nodes_bound` order affect the order of the returned array from `get_bound_child_nodes_to_bone()` and the `bound_children` property. So using `remove_unordered()` breaks compatibility I think.

**Review by lawnjelly — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by timothyqiu** [comment]
Cherry-picked for 3.5.2

**Pull request merged**